porch/func/internal/podevaluator.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/GoogleContainerTools/kpt/porch/func/evaluator"
+	pb "github.com/GoogleContainerTools/kpt/porch/func/evaluator"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+const (
+	defaultWrapperServerPort = 9446
+	volumeMountPath          = "/wrapper-server-tools"
+)
+
+type podEvaluator struct {
+	pb.UnimplementedFunctionEvaluatorServer
+
+	// kubeClient is the kubernetes client
+	kubeClient client.Client
+	// namespace holds the namespace where the executors run
+	namespace string
+	// wrapperServerImage is the image name of the wrapper server
+	wrapperServerImage string
+}
+
+func NewPodEvaluator(namespace, wrapperServerImage string) (pb.FunctionEvaluatorServer, error) {
+	restCfg, err := config.GetConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get rest config: %w", err)
+	}
+	cl, err := client.New(restCfg, client.Options{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create client: %w", err)
+	}
+	return &podEvaluator{
+		kubeClient:         cl,
+		namespace:          namespace,
+		wrapperServerImage: wrapperServerImage,
+	}, nil
+}
+
+func (pe *podEvaluator) EvaluateFunction(ctx context.Context, req *pb.EvaluateFunctionRequest) (*pb.EvaluateFunctionResponse, error) {
```

Comment:
this is nice! we'll need to add the optimizations but I am excited about this approach